src/extract/global_greedy_dag.rs

```diff
@@ -0,0 +1,193 @@
+    hash_cons: HashMap<Term, TermId>,
+}
+
+impl TermDag {
+    // Makes a new term using a node and children terms
+    // Correctly computes total_cost with sharing
+    // If this term contains itself, returns None
+    // If this term costs more than target, returns None
+    pub fn make(
+        &mut self,
+        node_id: NodeId,
+        node: &Node,
+        children: Vec<TermId>,
+        target: NotNan<f64>,
+    ) -> Option<TermId> {
+        let term = Term {
+            op: node.op.clone(),
+            children: children.clone(),
+        };
+
+        if let Some(id) = self.hash_cons.get(&term) {
+            return Some(*id);
+        }
+
+        let node_cost = node.cost;
+
+        if children.is_empty() {
+            let next_id = self.nodes.len();
+            self.nodes.push(term.clone());
+            self.info.push(TermInfo {
+                node: node_id,
+                eclass: node.eclass.clone(),
+                node_cost,
+                total_cost: node_cost,
+                reachable: [node.eclass.clone()].into_iter().collect(),
+                size: 1,
+            });
+            self.hash_cons.insert(term, next_id);
+            Some(next_id)
+        } else {
+            // check if children contains this node
+            for child in &children {
+                if self.info[*child].reachable.contains(&node.eclass) {
+                    return None;
+                }
+            }
+
+            let biggest_child = (0..children.len())
+                .max_by_key(|i| self.info[children[*i]].size)
+                .unwrap();
+
+            let mut cost = node_cost + self.total_cost(children[biggest_child]);
+            let mut reachable = Box::new(self.info[children[biggest_child]].reachable.clone());
+            let next_id = self.nodes.len();
+
+            for child in children.iter() {
+                if cost > target {
+                    return None;
+                }
+                let child_cost = self.get_cost(&mut reachable, *child);
+                cost += child_cost;
+            }
+
+            if cost > target {
+                return None;
+            }
+
+            *reachable = reachable.insert(node.eclass.clone());
+
+            self.info.push(TermInfo {
+                node: node_id,
+                node_cost,
+                eclass: node.eclass.clone(),
+                total_cost: cost,
+                reachable: *reachable,
+                size: 1 + children.iter().map(|c| self.info[*c].size).sum::<usize>(),
+            });
+            self.nodes.push(term.clone());
+            self.hash_cons.insert(term, next_id);
+            Some(next_id)
+        }
+    }
+
+    // Return a new term, like this one but making use of shared terms.
+    // Also return the cost of the new nodes.
+    fn get_cost(&self, shared: &mut Box<Reachable>, id: TermId) -> NotNan<f64> {
+        let eclass = self.info[id].eclass.clone();
+        if shared.contains(&eclass) {
```

Comment:
I think this is the key "why" for how this is better than greedy-dag: instead of doing an $O(n)$ union operation when you merge reachable sets, we "stop early" based on the shared egraph-level structure: if two nodes share a child somewhere down we don't continue unioning because we know the sets are the same. 

Is that right? If so, I think it's worth a comment :) (If not, I'd *definitely* appreciate a comment)

Reply:
Exactly right- I left a comment
// This is the key to why this algorithm is faster than greedy_dag.
// While doing the set union between reachable sets, we can stop early
// if we find a shared term.
// Since the term with `id` is shared, the reachable set of `id` will already
// be in `shared`.